dotnet/samples/HomeAutomation/Program.cs

```diff
@@ -0,0 +1,67 @@
+﻿/*
+ Copyright (c) Microsoft. All rights reserved.
+
+ Example application that uses Semantic Kernel.
+
+ Loads app configuration from:
+ - appsettings.json.
+ - appsettings.{Environment}.json.
+ - Secret Manager when the app runs in the "Development" environment (set through the DOTNET_ENVIRONMENT variable).
+ - Environment variables.
+ - Command-line arguments.
+*/
+
+using HomeAutomation.Options;
+using HomeAutomation.Plugins;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.OpenAI;
+
+namespace HomeAutomation;
+
+internal static class Program
+{
+    internal static async Task Main(string[] args)
+    {
+        HostApplicationBuilder builder = Host.CreateApplicationBuilder(args);
+
+        // Actual code to execute is found in Worker class
+        builder.Services.AddHostedService<Worker>();
+
+        // Get configuration
+        builder.Services.AddOptions<AzureOpenAiOptions>()
+                        .Bind(builder.Configuration.GetSection(nameof(AzureOpenAiOptions)))
+                        .ValidateDataAnnotations()
+                        .ValidateOnStart();
+
+        // Add plugins to include in all kernel instances here
+        builder.Services.AddSingleton<IEnumerable<KernelPlugin>>(sp =>
```

Comment:
Ben and I synced with Stephen, and he recommended that we could use the following approach to avoid the "magic" feeling of injection a KernelPluginCollection (especially if you have multiple kernels).

```
// Add plugins that can be used by kernels
// The plugins are added as singletons so that they can be used by multiple kernels
builder.Services.AddSingleton<MyTimePlugin>();
builder.Services.AddKeyedSingleton<MyLightPlugin>("OfficeLight");
builder.Services.AddKeyedSingleton<MyLightPlugin>("PorchLight", (sp, key) =>
{
    return new MyLightPlugin(turnedOn: false);
});

// Add a home automation kernel to the dependency injection container
builder.Services.AddKeyedTransient<Kernel>("HomeAutomationAgent", (sp, key) =>
{
    // Create a collection of plugins that the kernel will use
    KernelPluginCollection pluginCollection = new();
    pluginCollection.AddFromObject(sp.GetRequiredService<MyTimePlugin>());
    pluginCollection.AddFromObject(sp.GetRequiredKeyedService<MyLightPlugin>("OfficeLight"), "OfficeLight");
    pluginCollection.AddFromObject(sp.GetRequiredKeyedService<MyLightPlugin>("PorchLight"), "PorchLight");

    return new Kernel(sp, pluginCollection);
});
```

Reply:
Done